packages/language-service/src/lib-new/features/ls-st-import.ts

```diff
@@ -0,0 +1,390 @@
+                    addPathRelativeCompletions({
+                        completions,
+                        context,
+                        contextDirPath: context.fs.dirname(packageJsonPath),
+                        originFilePath,
+                        specifierPath: internalPath || './',
+                    });
+                }
+            }
+        } else {
+            // package names completions
+            const allowFullCompletions = pathBeforeCaret === '';
+            for (const packageName of Object.keys(packages)) {
+                let deltaStart = 0;
+                if (packageName.startsWith(pathBeforeCaret)) {
+                    deltaStart = -pathBeforeCaret.length;
+                } else if (!allowFullCompletions) {
+                    continue;
+                }
+                const label = packageName;
+                const detail = '';
+                const snippet = new Snippet(label);
+                completions.push(
+                    new Completion(
+                        label,
+                        detail,
+                        'a',
+                        snippet,
+                        range(context.getPosition(), { deltaStart })
+                    )
+                );
+            }
+        }
+    }
+}
+// naive check for an actual object
+function isObject(obj: any): obj is JSON {
+    return !!obj && !Array.isArray(obj) && typeof obj === 'object';
+}
+function getNodeModules({
+    context: { fs },
+    originDirPath,
+}: {
+    context: LangServiceContext;
+    originDirPath: string;
+}) {
+    const packages: Record<string, string> = {};
+    let dirPath = originDirPath;
+    let searching = true;
+    while (searching) {
+        // search packages
+        const nodeModulesPath = fs.join(dirPath, 'node_modules');
+        const stat = fs.statSync(nodeModulesPath, { throwIfNoEntry: false });
+        if (stat) {
+            const items = fs.readdirSync(nodeModulesPath, { withFileTypes: true });
+            for (const item of items) {
+                if (item.isDirectory()) {
+                    const packagePath = fs.join(nodeModulesPath, item.name);
+                    if (item.name[0] === '@') {
+                        const scopedItems = fs.readdirSync(packagePath, {
+                            withFileTypes: true,
+                        });
+                        for (const scopedItem of scopedItems) {
+                            const scopedPackageName = item.name + '/' + scopedItem.name;
+                            if (scopedItem.isDirectory() && !packages[scopedPackageName]) {
+                                packages[scopedPackageName] = fs.join(packagePath, scopedItem.name);
+                            }
+                        }
+                    } else if (!packages[item.name]) {
+                        packages[item.name] = packagePath;
+                    }
+                }
+            }
+        }
+        if (dirPath === '/') {
```

Comment:
Also, you're using `fs.join`, which had me wondering too.

Reply:
good catch! changed tests to run on os fs instead of memory fs and fixed issues.

 > Also, you're using fs.join, which had me wondering too.

using `IFileSystem` from `@file-services` that mashes fs and path together.